async/kafka/group/group.go

```diff
@@ -0,0 +1,179 @@
+package group
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"runtime"
+
+	"github.com/Shopify/sarama"
+	"github.com/beatlabs/patron/async"
+	"github.com/beatlabs/patron/async/kafka"
+	"github.com/beatlabs/patron/errors"
+	"github.com/beatlabs/patron/log"
+	"github.com/opentracing/opentracing-go"
+)
+
+// Factory definition of a consumer factory.
+type Factory struct {
+	name    string
+	group   string
+	topic   string
+	brokers []string
+	oo      []kafka.OptionFunc
+}
+
+// New constructor.
+func New(name, group, topic string, brokers []string, oo ...kafka.OptionFunc) (*Factory, error) {
+
+	if name == "" {
+		return nil, errors.New("name is required")
+	}
+
+	if group == "" {
+		return nil, errors.New("group is required")
+	}
+
+	if len(brokers) == 0 {
+		return nil, errors.New("provide at least one broker")
+	}
+
+	if topic == "" {
+		return nil, errors.New("topic is required")
+	}
+
+	return &Factory{name: name, group: group, topic: topic, brokers: brokers, oo: oo}, nil
+}
+
+// Create a new consumer.
+func (f *Factory) Create() (async.Consumer, error) {
+
+	config, err := kafka.SaramaConfig(f.name)
+
+	if err != nil {
+		return nil, err
+	}
+
+	cc := kafka.ConsumerConfig{
+		Brokers:      f.brokers,
+		Buffer:       0,
+		SaramaConfig: config,
+	}
+
+	c := &consumer{
+		topic:       f.topic,
+		group:       f.group,
+		traceTag:    opentracing.Tag{Key: "group", Value: f.group},
+		consumerCnf: cc,
+	}
+
+	for _, o := range f.oo {
+		err = o(c)
+		if err != nil {
+			return nil, fmt.Errorf("Could not apply OptionFunc '%v' to consumer : %v", runtime.FuncForPC(reflect.ValueOf(o).Pointer()).Name(), err)
```

Comment:
We should try to avoid using the reflect package except for cases where we cannot avoid it. The OptionFunc error is enough to identify what's wrong.